packages/core/src/browser/source-tree/source-tree.ts

```diff
@@ -41,50 +43,56 @@ export class SourceTree extends TreeImpl {
         return parent.element.getElements();
     }
 
-    protected toNode(element: TreeElement, index: number, parent: TreeElementNodeParent): TreeElementNode {
-        const id = element.id ? String(element.id) : (parent.id + ':' + index);
-        const name = id;
-        const existing = this.getNode(id);
-        const updated = existing && <TreeElementNode>Object.assign(existing, { element, parent });
+    /**
+     * Convert a parent's elements into `TreeElementNode` or `CompositeTreeElementNode`.
+     *
+     * If a node already exists in this tree, we'll mutate it rather than create a new updated one.
+     */
+    protected toNode(element: TreeElement, index: number, parent: TreeElementNodeParent): TreeElementNode | CompositeTreeElementNode {
+        const id: string = element.id === undefined ? `${parent.id}:${index}` : String(element.id);
+        // The base `TreeImpl` class expects to store simple `TreeNode` instances,
+        // but we know for a fact that we handle `TreeElementNode` in this `SourceTree` class.
+        // TODO: `Tree` should be generic and require something like `T extends TreeNode`?
+        const existing = this.getNode(id) as TreeElementNode | CompositeTreeElementNode | undefined;
+        if (existing) {
+            existing.element = element;
+            existing.parent = parent;
+        }
         if (CompositeTreeElement.hasElements(element)) {
-            if (updated) {
-                if (!ExpandableTreeNode.is(updated)) {
-                    Object.assign(updated, { expanded: false });
+            if (existing) {
+                if (!ExpandableTreeNode.is(existing)) {
+                    (existing as any).expanded = false;
```

Comment:
TS seems happy with these `Object.assign` calls. Why do you prefer this style?

Reply:
Because TS checks nada when doing `Object.assign`... It's worse that using `any` and relies on useless reflection.

Reply:
In that case, I think the `Partial<ExpandableTreeNode>` trick will work here, as well, avoiding the `any`.

Reply:
I updated the code with `Partial` like you suggested. We're still kinda lying to the compiler, but without `any`s now.